src/reqstool/common/validators/lifecycle_validator.py

```diff
@@ -0,0 +1,54 @@
+# Copyright © LFV
+
+import logging
+from typing import Union
+from reqstool.common.dataclasses.lifecycle import LIFECYCLESTATE
+from reqstool.common.dataclasses.urn_id import UrnId
+from reqstool.models.annotations import AnnotationData
+from reqstool.models.combined_indexed_dataset import CombinedIndexedDataset
+from reqstool.models.requirements import RequirementData
+from reqstool.models.svcs import SVCData
+from reqstool_python_decorators.decorators.decorators import Requirements
+
+
+@Requirements("REQ_037", "REQ_038")
+class LifecycleValidator:
+    """
+    Logs warnings if any requirement or SVC is used despite being marked deprecated or obsolete.
+    """
+
+    def __init__(self, cid: CombinedIndexedDataset):
+        self._cid = cid
+
+        self._validate()
+
+    def _validate(self):
+        self._check_defunct_annotations(self._cid.annotations_impls, self._cid.requirements)
+        self._check_defunct_annotations(self._cid.annotations_tests, self._cid.svcs)
+        self._check_defunct_references(self._cid.mvrs_from_svc, self._cid.svcs)
+        self._check_defunct_references(self._cid.svcs_from_req, self._cid.requirements)
+
+    def _check_defunct_annotations(
+        self,
+        annotations: dict[UrnId, list[AnnotationData]],
+        in_collection: dict[UrnId, Union[RequirementData, SVCData]],
+    ):
+        for urn in annotations:
+            if urn not in in_collection:
+                continue
+            state = in_collection[urn].lifecycle.state
+            if state == LIFECYCLESTATE.DEPRECATED or LIFECYCLESTATE.OBSOLETE:
+                logging.debug(f"{state.value}: {urn}")
+
+    def _check_defunct_references(
+        self, references: dict[UrnId, list[UrnId]], ref_collection: dict[UrnId, Union[SVCData, RequirementData]]
+    ):
+        for ref_id, related_ids in references.items():
+            if ref_id not in ref_collection:
+                continue
+            state = ref_collection[ref_id].lifecycle.state
+            if state == LIFECYCLESTATE.DEPRECATED or LIFECYCLESTATE.OBSOLETE:
+                logging.debug(f"{state.value}: {self._format_list(related_ids)}")
```

Comment:
Are these logging.debug intended for developers and/or users?

The PR description said "A warning is shown during parsing if a requirement or SVC marked as deprecated or obsolete is used."

So, if this is for a user then this log level is not sufficient to satisfy the PR description (I didn't see anything about this in the requirements for reqstool, perhaps worth adding?).

I'm not following entirely here what ref_collection is. Are you logging here? For each requirement that is depracted or obsolete any requirement or svc that is referencing it is logged? And the ID is the UrnId?

If you can provide a screenshot that'll be great.

Reply:
Yes of course it should not be debug, don't know how I forgot to change that.  
The variable name have been updated to hopefully be more clear.

This is what the output looks like:
![image](https://github.com/luftfartsverket/reqstool-client/assets/142237585/9b848097-b319-4da4-a43c-a5e63e102d8f)

Reply:
Ok. Perhaps an output that is more like a warning for users rather than a debug message for developer.

E.g.

"Requirement `ms-101:REQ_101` is used despite being deprecated."

and so on.

Also, what has priority (is most important info for the user) REQ/SVC vs deprecated/obsolete?
Think GROUP BY and ORDER BY?

Reply:
It now looks like this:

![image](https://github.com/luftfartsverket/reqstool-client/assets/142237585/7aa15c31-5f8f-4484-b414-854413952fc3)

Ordered in descending order by severity (obsolete first)